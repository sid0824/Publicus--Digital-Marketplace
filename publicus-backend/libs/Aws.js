import AWS from 'aws-sdk';
import {AccessKeyId, SecretAccessKey} from "../config/ConfigValues";

export const s3 = new AWS.S3({
  region: 'ap-southeast-2',
  accessKeyId: AccessKeyId,
  secretAccessKey: SecretAccessKey
});

export const ses = new AWS.SES({
  region: 'us-west-2'
});