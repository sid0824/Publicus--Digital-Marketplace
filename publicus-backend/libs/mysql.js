import mysql from 'mysql';
import {MySqlHost, MySqlPassword, MySqlUser} from "../config/ConfigValues";

//mysql connection to database
const connection = mysql.createConnection({
    host: MySqlHost,
    user: MySqlUser,
    password: MySqlPassword,
    database: "survey",
    multipleStatements: true
});

export const runQuery = async (query) => {
    return new Promise((resolve, reject) => {
        connection.query(query, (error, results) => {
            if(error) {
                reject(error);
            } else {
                resolve(results);
            }
        });
    });
};