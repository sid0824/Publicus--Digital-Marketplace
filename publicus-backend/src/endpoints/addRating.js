import {failure, success} from "../util/ResponseUtil";
import {runQuery} from "../../libs/mysql";

// Query to get a users current rating for a survey
const getExistingRatingQuery = (userId, surveyId) => {
  return `SELECT * FROM Ratings 
          WHERE userId = '${userId}' AND surveyId = '${surveyId}'`;
};

// Query to add a new rating for a user
const addNewRatingQuery = (userId, surveyId, newRating) => {
  return `INSERT INTO Ratings (userId, surveyId, rating)`
       + `VALUES ( '${userId}', '${surveyId}', ${newRating} )`;
};

// Query to update the rating for a user
const updateExistingRatingQuery = (userId, surveyId, newRating) => {
  return `UPDATE Ratings 
          SET rating = ${newRating} 
          WHERE userId = '${userId}' AND surveyId = '${surveyId}'`;
};

export async function main(event, context) {
  try {
    const body = JSON.parse(event.body);
    const { surveyId } = event.pathParameters;
    const userId = event.requestContext.identity.cognitoIdentityId;

    // Get current rating
    const currentRating = await runQuery(getExistingRatingQuery(userId, surveyId));

    // Update if a rating exists
    if (currentRating.length) {
      await runQuery(updateExistingRatingQuery(userId, surveyId, body.rating));
    } else {
      // Add if a rating doesnt exist
      await runQuery(addNewRatingQuery(userId, surveyId, body.rating));
    }
    return success();
  } catch (err) {
    console.error(err);
    return failure();
  }

}