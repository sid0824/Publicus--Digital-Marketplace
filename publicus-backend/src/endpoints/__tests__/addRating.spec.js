import { main } from '../addRating';
import { runQuery } from "../../../libs/mysql";

jest.mock("../../../libs/mysql");

const mockEvent = {
  requestContext: {
    identity: {
      cognitoIdentityId: 'MOCK_COGNITO_ID'
    },
  },
  body: JSON.stringify({
    rating: 2
  }),
  pathParameters: {
    surveyId: 'MOCK_SURVEY_ID'
  }
};

describe('addRating', () => {

  beforeEach(() => {

  });

  afterAll(() => {
    runQuery.mockReset();
  });

  test(`should respond with failure when an error occurs`, async () => {
    runQuery.mockReset();
    runQuery.mockImplementation(() => {
      throw new Error('ERROR')
    });
    const response = await main(mockEvent);
    expect(response).toMatchSnapshot();
  });

  test('should update a users rating for a survey', async () => {
    runQuery
      .mockResolvedValueOnce([{ rating: 4 }])
      .mockResolvedValue(null);
    const response = await main(mockEvent);
    expect(runQuery.mock.calls[0][0]).toMatchSnapshot();
    expect(runQuery.mock.calls[1][0]).toMatchSnapshot();
    expect(response).toMatchSnapshot();
  });


  test('should add a new rating for a users survey', async () => {
    runQuery
      .mockResolvedValueOnce([])
      .mockResolvedValue(null);
    const response = await main(mockEvent);
    expect(runQuery.mock.calls[0][0]).toMatchSnapshot();
    expect(runQuery.mock.calls[1][0]).toMatchSnapshot();
    expect(response).toMatchSnapshot();
  })

});