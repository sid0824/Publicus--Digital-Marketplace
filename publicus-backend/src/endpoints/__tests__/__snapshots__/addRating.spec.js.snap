// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addRating should add a new rating for a users survey 1`] = `
"SELECT * FROM Ratings 
          WHERE userId = 'MOCK_COGNITO_ID' AND surveyId = 'MOCK_SURVEY_ID'"
`;

exports[`addRating should add a new rating for a users survey 2`] = `
"SELECT * FROM Ratings 
          WHERE userId = 'MOCK_COGNITO_ID' AND surveyId = 'MOCK_SURVEY_ID'"
`;

exports[`addRating should add a new rating for a users survey 3`] = `
Object {
  "body": undefined,
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
  },
  "statusCode": 200,
}
`;

exports[`addRating should respond with failure when an error occurs 1`] = `
Object {
  "body": undefined,
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
  },
  "statusCode": 500,
}
`;

exports[`addRating should update a users rating for a survey 1`] = `
"SELECT * FROM Ratings 
          WHERE userId = 'MOCK_COGNITO_ID' AND surveyId = 'MOCK_SURVEY_ID'"
`;

exports[`addRating should update a users rating for a survey 2`] = `
"SELECT * FROM Ratings 
          WHERE userId = 'MOCK_COGNITO_ID' AND surveyId = 'MOCK_SURVEY_ID'"
`;

exports[`addRating should update a users rating for a survey 3`] = `
Object {
  "body": undefined,
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
  },
  "statusCode": 200,
}
`;
